### DISCUSSION TITLE
CAN bus interface

### DISCUSSION THREAD

**Discussion by Shivaraj199** [polarfire-soc/polarfire-soc-documentation#130] [Linux drivers for PolarFire Soc] [answered]
hello,
I am trying to use the classic can driver modules (can0 & can1) in Linux on the icicle kit. I am using the yocto BSP. I have enabled the can configurations in  /recipes-kernel/linux/files/icicle-kit-es-amp directory and include the file in recipe.
       CAN BUS  
       CONFIG_CAN=y
       CONFIG_CAN_RAW=y
       CONFIG_CAN_BCM=y
       CONFIG_CAN_GW=y
       CONFIG_CAN_GS_USB=y
       CONFIG_CAN_CALC_BITTIMING=y
       CONFIG_CAN_VCAN=y
       CONFIG_CAN_DEV=y
       CONFIG_CAN_MCP251X=y
With menuconfig i can confirm  bus controller is enabled, Also i changed the dtsi file in the /kernel-source/arch/riscv/boot/dts/microchip to enable can0 and can1

                 can0: can@2010c000 {
			compatible = "microchip,mpfs-can";
			reg = <0x0 0x2010c000 0x0 0x1000>;
			clocks = <&clkcfg CLK_CAN0>;
			interrupt-parent = <&plic>;
			interrupts = <56>;
			status = "okay";
	        };

		can1: can@2010d000 {
			compatible = "microchip,mpfs-can";
			reg = <0x0 0x2010d000 0x0 0x1000>;
			clocks = <&clkcfg CLK_CAN1>;
			interrupt-parent = <&plic>;
			interrupts = <57>;
			status = "okay";
		};

When i build the image and flash it in the icicle kit I can see that CAN_RAW is enable with dmesg, but when i try to add can0 i get an error  "Cannot find device "can0"
Can anyone please tell what i am missing out or to enable can0 & can1 in the kernel?

**Comment by Cyril-Jean** [comment]
Please follow the steps in the issue template, this issue should be filed as a discussion item.

**Comment by mcnamarad1971** [comment, marked as answer]
I'm thinking that you are trying to use the socketCAN interface? If so, that is currently not supported on PolarFire SoC.  We do have plans to implement it but no delivery date set yet.  

Right now, we advise people to use CAN from userspace; there is an example in /opt/microchip that shows how to do this.

**Reply by Shivaraj199** [reply]
Hello @mcnamarad1971 thankyou for the  details, I am changing my src code to use the UIO driver interface.

**Comment by Shivaraj199** [comment]


**Reply by ConchuOD** [reply]
>Is it possible if try to manually load the socketcan modules in the linux kenrel i can it? If yes then can try that approach.

Of course. If you have a socket can driver written (or want to use the socket can subsystem modules) you can load them in the standard ways.

**Reply by Shivaraj199** [reply]
@ConchuOD  Thanks,
Yes I am writing the driver code, The driver is only dependent on socketcan and standard libs so it should work on all systems that support socketcan as far as I can tell
I will crosscompile the driver to generate .ko and try to insert it using insmod.
I was bit confused because @mcnamarad1971 replied that socketcan is not supported by Polarfire SOC.

**Reply by ConchuOD** [reply]
There is no problem with socket can on PolarFire SoC other than that we have no yet written a socket can driver for it!
If you have written your own driver, it should be fine.
